lib/compile/apply-plugins.js

```diff
@@ -0,0 +1,17 @@
+module.exports = function applyPlugins (plugins, ev /* args... */) {
+  if (!ev) {
+    throw new Error("Must provide event type for plugin application.");
+  }
+
+  const contextArgs = Array.prototype.slice.call(arguments, 3);
+  let result = arguments[2];
+
+  plugins.forEach(plugin => {
```

Comment:
Not sure if this is to support other things in the code but it seems like you have the wrong data structure here if you have to loop over it at every lifecycle step to find the valid plugins. Not that this is a perf bottleneck but you could at least memoize the value.